Unit tests for viewRecords

hannespetur opened the pull request:
Pull request within a pull request. A pullception (copyright under consideration).

Several unit tests for `viewRecords` added. This change also adds a few additional tests for the `jumpToRegion` function, which makes sure the correct read is "jumped" to when the function is used.

Best,
Hannes

hannespetur commented:
@smehringer the tests for viewRecords are not passing right now, but I am not sure why :astonished:  I will try run `gdb` to see what is going on..

I will rebase my branch if you make any changes so don't worry about creating conflicts.

Best wishes,
Hannes

hannespetur commented:
Nevermind, the failed test was due to a bug in one of the tests. Do you think these unit tests will cover enough @smehringer ?

smehringer commented on include/seqan/bam_io/bam_index_bai.h:
```diff
@@ -61,6 +61,9 @@
 #ifndef INCLUDE_SEQAN_BAM_IO_BAM_INDEX_BAI_H_
 #define INCLUDE_SEQAN_BAM_IO_BAM_INDEX_BAI_H_
 
+#include <cstdint>
```
why this?

hannespetur replied:
Oops, I forgot to remove those... I have been trying out different IDEs lately and one gave me all these "Undefined type 'uint32_t' blahblah" errors without these two includes. I will remove them.

smehringer commented on tests/bam_io/test_bam_index.h:
```diff
@@ -84,15 +84,30 @@ SEQAN_DEFINE_TEST(test_bam_io_bam_index_open)
+    SEQAN_ASSERT_NOT(jumpToRegion(bamFile, hasAlignments, 1, 1, 10, baiIndex));
+    SEQAN_ASSERT_NOT_MSG(hasAlignments, "No alignments should be on contig 1.");
+
+    SEQAN_ASSERT(jumpToRegion(bamFile, hasAlignments, 0, 4, 100, baiIndex));
```
The tests always have a read starting exactly at the beginning of the region. Maybe we add one more read at pos... 42? and jump to region `[4,100]` expecting success.

hannespetur replied:
Good point. There are some other tests that depend on the `small.bam` file that I don't want to break though.. I think I will rewrite the `jumpToRegion` tests using `ex1.bam` instead so I can more easily cover other cases.

smehringer commented on tests/bam_io/test_bam_index.h:
```diff
@@ -109,5 +124,79 @@ SEQAN_DEFINE_TEST(test_bam_io_bam_index_save)
+
+    std::vector<BamAlignmentRecord> records;
+
+    viewRecords(records, bamFile, baiFile, 0, 1, 0);
```
Nice test case :+1: . we should add this for jumpToRegion

smehringer replied:
we should also add one with the region `[0, 100]` where it is clearly not 1-based.

hannespetur replied:
Agreed. But the region `[0, 100]` should throw an assertion failure, is it possible to check for that?

smehringer commented on tests/bam_io/test_bam_index.h:
```diff
@@ -109,5 +124,79 @@ SEQAN_DEFINE_TEST(test_bam_io_bam_index_save)
+    viewRecords(records, bamFile, baiFile, 0, 2000, 2000);
+    SEQAN_ASSERT_MSG(length(records) == 0, "No reads expected in region seq1:2000-2000");
+    viewRecords(records, bamFile, baiFile, 0, 2000, 100000);
+    SEQAN_ASSERT_MSG(length(records) == 0, "No reads expected in region seq1:2000-100000");
```
what is the difference to the test case before?

hannespetur replied:
None, really :smile:

smehringer commented:
PS: better make that copyright fast :running_man:  :dash:

hannespetur commented:
Thank you for a fast response and a good review @smehringer! I think I have addressed all your remarks except for one that I wasn't sure how to do in SeqAn (add test with region `[0, 100]` that should fail).

smehringer commented:
There is the `SEQAN_TEST_EXCEPTION_WITH_MESSAGE(...)` macro :)
I will add it, since I noticed that I only included a debug assertion but did not throw anything so far :innocent:

smehringer commented on include/seqan/bam_io/bam_index_bai.h:
```diff
@@ -61,6 +61,7 @@
 #ifndef INCLUDE_SEQAN_BAM_IO_BAM_INDEX_BAI_H_
 #define INCLUDE_SEQAN_BAM_IO_BAM_INDEX_BAI_H_
 
+
```
:nail_care:

hannespetur replied:
The file is clearly unreadable without that newline 😃

smehringer commented:
@hannespetur Do you think we should also throw an assertion if the specified region end position is larger than the actual contig size? I can check for this.. but i don't know if an exception should be thrown or if the region end should simply be put to the length of the contig. (I have a developer in mind doesn't check for region boundaries based on an user input. So the user just wants a tool to run through, but because of the careless developer the programm throws an assertion... :thinking: )

hannespetur commented:
IMO it is nicer if the region end should be put as the length of the contig instead of throwing an exception. `samtools view` handles it like that that and I think it is fine. I think it is alright not to throw an exception when the user's intent is clear.

I also have noticed that `samtools view` parses regions starting at position 0 without an error, e.g. if passed chr21:0-10000000 it just assumes the user means chr21:1-10000000. Maybe that case should not throw either.

smehringer commented:
Thanks for checking samtools! I agree we should not throw in this case but I will add debug assertions for both cases. It's more user friendly to not let the program abort.

hannespetur commented:
Great, many thanks!
